core/shared/src/main/scala/spire/math/package.scala

```diff
@@ -403,10 +403,126 @@ package object math {
   final def hypot[@sp(Float, Double) A](x: A, y: A)
     (implicit f: Field[A], n: NRoot[A], o: Order[A]): A = {
     import spire.implicits._
-    if (x > y) x.abs * (1 + (y/x)**2).sqrt
-    else y.abs * (1 + (x/y)**2).sqrt
+    def abs(n: A): A = if (n < f.zero) -n else n
+    if (x > y) abs(x) * (1 + (y/x)**2).sqrt
+    else abs(y) * (1 + (x/y)**2).sqrt
   }
 
+  // BigInt
+  /**
+   * This will return the largest integer that meets some criteria. Specifically,
+   * if we're looking for some integer `x` and `f(x')` is guaranteed to return
+   * `true` iff `x' <= x`, then this will return `x`.
+   *
+   * This can be used, for example, to find an integer `x` s.t. 
+   * `x * x < y < (x+1)*(x+1)`, by using `intSearch(x => x * x <= y)`.
+   */
+  private def intSearch(f: Int => Boolean): Int = {
+    val ceil = (0 until 32) find (i => !f(1 << i)) getOrElse 33
+    if (ceil == 0) {
+      0
+    } else {
+      (0 /: ((ceil - 1) to 0 by -1)) { (x, i) =>
+        val y = x | (1 << i)
+        if (f(y)) y else x
+      }
+    }
+  }
+
+
+  /**
+   * Returns the digits to the right of the decimal point of `x / y` in base
+   * `r` if x < y.
+   */
+  private def decDiv(x: BigInt, y: BigInt, r: Int): Stream[BigInt] = {
+    val expanded = x * r
+    val quot = expanded / y
+    val rem = expanded - (quot * y)
+
+    if (rem == 0) {
+      Stream.cons(quot, Stream.empty)
+    } else {
+      Stream.cons(quot, decDiv(rem, y, r))
+    }
+  }
+
+  /** Returns the digits of `x` in base `r`. */
+  private def digitize(x: BigInt, r: Int, prev: List[Int] = Nil): List[Int] =
+    if (x == 0) prev else digitize(x / r, r, (x % r).toInt :: prev)
+
+
+  /** Converts a list of digits in base `r` to a `BigInt`. */
+  private def undigitize(digits: Seq[Int], r: Int): BigInt =
+    (BigInt(0) /: digits)(_ * r + _)
+
+
+  // 1 billion: because it's the largest positive Int power of 10.
+  private val radix = 1000000000
+
+  /**
+   * An implementation of the shifting n-th root algorithm for BigDecimal. For
+   * the BigDecimal a, this is guaranteed to be accurate up to the precision
+   * specified in ctxt.
+   *
+   * See http://en.wikipedia.org/wiki/Shifting_nth_root_algorithm
+   *
+   * @param a A (positive if k % 2 == 0) `BigDecimal`.
+   * @param k A positive `Int` greater than 1.
+   * @param ctxt The `MathContext` to bound the precision of the result.
+   *
+   * returns A `BigDecimal` approximation to the `k`-th root of `a`.
+   */
+  def nroot(a: BigDecimal, k: Int, ctxt: MathContext): BigDecimal =
```

Comment:
Nothing major here (and feel free to do nothing), but it may be that we can just delete this code entirely and use `Algebraic.nroot(a.bigDecimal, k, mc)` instead, for the time being. If you actually use `Algebraic(a).nroot(k).toBigDecimal(mc)` instead, you'll get exact rounding, but it'll be a bit slower and probably not worth it, given you can't expect much correctness with BigDecimal.